src/runtime/gpu_context_common.h

```diff
@@ -0,0 +1,192 @@
+    static constexpr uint32_t kDeletedId{1};
+
+    uint32_t unique_id{2};  // zero is an invalid id
+
+    static ALWAYS_INLINE uintptr_t kernel_hash(ContextT context, uint32_t id, uint32_t bits) {
+        uintptr_t addr = (uintptr_t)context + id;
+        // Fibonacci hashing. The golden ratio is 1.9E3779B97F4A7C15F39...
+        // in hexadecimal.
+        if (sizeof(uintptr_t) >= 8) {
+            return (addr * (uintptr_t)0x9E3779B97F4A7C15) >> (64 - bits);
+        } else {
+            return (addr * (uintptr_t)0x9E3779B9) >> (32 - bits);
+        }
+    }
+
+    bool insert(ContextT context, uint32_t id, ModuleStateT module_state) {
+        if (log2_compilations_size == 0) {
+            if (!resize_table(kInitialTableBits)) {
+                return false;
+            }
+        }
+        if ((count + 1) > (1 << log2_compilations_size) * kLoadFactor) {
+            if (!resize_table(log2_compilations_size + 1)) {
+                return false;
+            }
+        }
+        count += 1;
+        uintptr_t index = kernel_hash(context, id, log2_compilations_size);
+        for (int i = 0; i < (1 << log2_compilations_size); i++) {
+            uintptr_t effective_index = (index + i) & ((1 << log2_compilations_size) - 1);
+            if (compilations[effective_index].kernel_id <= kDeletedId) {
+                compilations[effective_index].context = context;
+                compilations[effective_index].module_state = module_state;
+                compilations[effective_index].kernel_id = id;
+                return true;
+            }
+        }
+        // This is a logic error that should never occur. It means the table is
+        // full, but it should have been resized.
+        halide_assert(nullptr, false);
+        return false;
+    }
+
+    bool find_internal(ContextT context, uint32_t id, ModuleStateT *&module_state) {
+        if (log2_compilations_size == 0) {
+            return false;
+        }
+        uintptr_t index = kernel_hash(context, id, log2_compilations_size);
+        for (int i = 0; i < (1 << log2_compilations_size); i++) {
+            uintptr_t effective_index = (index + i) & ((1 << log2_compilations_size) - 1);
+
+            if (compilations[effective_index].kernel_id == kInvalidId) {
+                return false;
+            }
+            if (compilations[effective_index].context == context &&
+                compilations[effective_index].kernel_id == id) {
+                module_state = &compilations[effective_index].module_state;
+                return true;
+            }
+        }
+        return false;
+    }
+
+    bool lookup(ContextT context, void *state_ptr, ModuleStateT &module_state) {
+        ScopedMutexLock lock_guard(&mutex);
+        uint32_t id = (uint32_t)(uintptr_t)state_ptr;
+        ModuleStateT *mod_ptr;
+        if (find_internal(context, id, mod_ptr)) {
+            module_state = *mod_ptr;
+            return true;
+        }
```

Comment:
Might be safer to reliably set module_state = null upon failure, so caller can't mistakenly use garbage?

Reply:
The code is templated in a way where it is not entirely guaranteed a nullptr value fits the type here. Ideally it would use an optional return, but we don't have one handy.